### DISCUSSION TITLE
Autotagging items in library

### DISCUSSION THREAD

**Discussion by Halem998** [papis/papis#604] [Ideas]
As was brought up in issue #564 , tagging items seems crucial in organizing your stuff.
Now, I'm a lazy person, I refrain from tagging 2000 items in retrospect.
What would be cool:

A code / plugin for `autotagging` based on OCR / contextual search (maybe in a GPT direction)  + tags which are already have been generated from metadata upon adding to the database ...

Maybe this already exists, but I'm not aware of this.
Certainly I'm not capable of writing this (yet).

Any thoughts?

**Comment by Halem998** [comment]
https://github.com/bellecp/fast-p

a code like the above works pretty well.
But it's not quite enough to get where I would imagine ...

**Reply by jghauser** [reply]
I use something similar to full-text search all my docs using [ripgrep-all](https://github.com/phiresky/ripgrep-all) and [fzf](https://github.com/junegunn/fzf):

```bash
#!/usr/bin/env bash

RG_PREFIX="rga --files-with-matches --hidden --glob '!.git'"
file="$(
  FZF_DEFAULT_COMMAND="$RG_PREFIX '$1'" \
    fzf --sort --preview="[[ ! -z {} ]] && rga --pretty --context 5 {q} {}" \
      --phony -q "$1" \
      --bind "change:reload:$RG_PREFIX {q}" \
      --preview-window="70%:wrap"
)" &&
echo "opening $file" &&
xdg-open "$file"
```

When it comes to tagging, I've started doing the slow manual thing of adding them whenever I add a document. That is somewhat made easier by [papis.nvim](https://github.com/jghauser/papis.nvim) auto-completing tags when editing the `info.yaml` files. It would definitely be nice if some of this could be automated, though I fear my own classification scheme might be too idiosyncratic for a machine haha.

**Reply by jghauser** [reply]
I've actually been thinking about integrating this full-text search ability into papis.nvim, but haven't gotten around to it yet.

**Reply by Halem998** [reply]
Ah, that code is interesting! thanks for the input :))

**Reply by Halem998** [reply]
About the 'idiosyncracy': I totally agree with yout that in the end, there must be some individual structure to all of our way of tagging. But something like GPT could at least give rudimentary proposals at least? hm ...

Maybe I'll just vim through all the single documents in a day and drink lots and lots of coffee ;)

**Reply by alejandrogallo** [reply]
a GPT thing would be cool, but somehow I think that having a coherent and systematic tagging system is quite better. However I wonder how well GPT4 performs given the prompt: give me 4 tags describing the content of the following publication.

**Reply by alejandrogallo** [reply]
one alternative would also be to create a public papis library on github with hundreds of thousands of publications from papis users all around, and contribute the curated tags and notes that papis users have written about them. I have been thinking about this for some years now and as long as we do not commit the pdfs that would be quite safe. Maybe we have enough momentum now to start even thinking about such a thing.

**Reply by jghauser** [reply]
I think I might actually underestimate GPT, especially if we feed in some existing categories used in the one's library. As in, "Please add 4 tags for each documents and follow the categorisation scheme used so far in the library". Just dreaming here... :D

This could of course be very usefully combined with something like what @alejandrogallo is suggesting (especially for the docs that don't yet exist in others' libraries)!

**Reply by Halem998** [reply]
@alejandrogallo 

I think that youre right, I also think these aspects dont exclude each other necessarily > 
Actually, I simply often **don't know ** how to tag them, and would love to have some rudimentary autotag, but still be able to control that  to a certain degree ...

https://github.com/openai/chatgpt-retrieval-plugin

**Reply by Halem998** [reply]
Great to see your ideas floating around :)
